edb/server/compiler/compiler.py

```diff
@@ -636,136 +644,151 @@ def _compile_ql_query(
 
         sql_bytes = sql_text.encode(defines.EDGEDB_ENCODING)
 
-        if single_stmt_mode:
-            if native_out_format:
-                out_type_data, out_type_id = sertypes.TypeSerializer.describe(
-                    ir.schema, ir.stype,
-                    ir.view_shapes, ir.view_shapes_metadata,
-                    inline_typenames=ctx.inline_typenames,
-                    protocol_version=ctx.protocol_version)
-            else:
-                out_type_data, out_type_id = \
-                    sertypes.TypeSerializer.describe_json()
-
-            in_type_args = None
-            params: list[tuple[str, s_obj.Object, bool]] = []
-            has_named_params = False
-
-            if ir.params:
-                first_param = next(iter(ir.params))
-                has_named_params = not first_param.name.isdecimal()
-                if (src := ctx.source) is not None:
-                    first_extracted = src.first_extra()
-                else:
-                    first_extracted = None
+        in_type_args = None
+        params: list[tuple[str, s_obj.Object, bool]] = []
+        has_named_params = False
+        if ir.params:
+            params, in_type_args = self._extract_params(
+                ir.params, argmap=argmap, script_info=script_info,
+                schema=ir.schema, ctx=ctx)
+
+        globals = None
+        if ir.globals:
+            globals = [
+                (str(glob.global_name), glob.has_present_arg)
+                for glob in ir.globals
+            ]
 
-                if first_extracted is not None:
-                    user_params = first_extracted
-                else:
-                    user_params = len(ir.params)
+        if (
+            ctx.output_format is enums.OutputFormat.NONE or
+            ctx.stmt_mode is not enums.CompileStatementMode.SINGLE
+        ):
+            out_type_id = sertypes.NULL_TYPE_ID
+            out_type_data = sertypes.NULL_TYPE_DESC
+            result_cardinality = enums.Cardinality.NO_RESULT
+        elif ctx.output_format is enums.OutputFormat.BINARY:
+            out_type_data, out_type_id = sertypes.TypeSerializer.describe(
+                ir.schema, ir.stype,
+                ir.view_shapes, ir.view_shapes_metadata,
+                inline_typenames=ctx.inline_typenames,
+                protocol_version=ctx.protocol_version)
+        else:
+            out_type_data, out_type_id = \
+                sertypes.TypeSerializer.describe_json()
+
+        if ctx.protocol_version >= (0, 12):
+            in_type_data, in_type_id = \
+                sertypes.TypeSerializer.describe_params(
+                    schema=ir.schema,
+                    params=params,
+                    protocol_version=ctx.protocol_version,
+                )
+        else:
+            # Legacy protocol support
```

Comment:
Drop this?  We shouldn't be supporting 0.11 and below anymore.

Reply:
I think this is for restoring pre-0.12 dumps - I'll complete the comment here.